### ISSUE TITLE
I can't understand the calculation of Covariance of a set of points in fcl/BVH/BVH_utility.cpp

### ISSUE THREAD

**Issue by betterfan** [flexible-collision-library/fcl#268]
anyone can clarify it? thank you ...^-^
void getCovariance(Vec3f* ps, Vec3f* ps2, Triangle* ts, unsigned int* indices, int n, Matrix3f& M)
{
  Vec3f S1;
  Vec3f S2[3];

  if(ts)
  {
    for(int i = 0; i < n; ++i)
    {
      const Triangle& t = (indices) ? ts[indices[i]] : ts[i];

      const Vec3f& p1 = ps[t[0]];
      const Vec3f& p2 = ps[t[1]];
      const Vec3f& p3 = ps[t[2]];

      S1[0] += (p1[0] + p2[0] + p3[0]);
      S1[1] += (p1[1] + p2[1] + p3[1]);
      S1[2] += (p1[2] + p2[2] + p3[2]);
      S2[0][0] += (p1[0] * p1[0] + p2[0] * p2[0] + p3[0] * p3[0]);
      S2[1][1] += (p1[1] * p1[1] + p2[1] * p2[1] + p3[1] * p3[1]);
      S2[2][2] += (p1[2] * p1[2] + p2[2] * p2[2] + p3[2] * p3[2]);
      S2[0][1] += (p1[0] * p1[1] + p2[0] * p2[1] + p3[0] * p3[1]);
      S2[0][2] += (p1[0] * p1[2] + p2[0] * p2[2] + p3[0] * p3[2]);
      S2[1][2] += (p1[1] * p1[2] + p2[1] * p2[2] + p3[1] * p3[2]);

      if(ps2)
      {
        const Vec3f& p1 = ps2[t[0]];
        const Vec3f& p2 = ps2[t[1]];
        const Vec3f& p3 = ps2[t[2]];

        S1[0] += (p1[0] + p2[0] + p3[0]);
        S1[1] += (p1[1] + p2[1] + p3[1]);
        S1[2] += (p1[2] + p2[2] + p3[2]);

        S2[0][0] += (p1[0] * p1[0] + p2[0] * p2[0] + p3[0] * p3[0]);
        S2[1][1] += (p1[1] * p1[1] + p2[1] * p2[1] + p3[1] * p3[1]);
        S2[2][2] += (p1[2] * p1[2] + p2[2] * p2[2] + p3[2] * p3[2]);
        S2[0][1] += (p1[0] * p1[1] + p2[0] * p2[1] + p3[0] * p3[1]);
        S2[0][2] += (p1[0] * p1[2] + p2[0] * p2[2] + p3[0] * p3[2]);
        S2[1][2] += (p1[1] * p1[2] + p2[1] * p2[2] + p3[1] * p3[2]);
      }
    }
  }
  else
  {
    for(int i = 0; i < n; ++i)
    {
      const Vec3f& p = (indices) ? ps[indices[i]] : ps[i];
      S1 += p;
      S2[0][0] += (p[0] * p[0]);
      S2[1][1] += (p[1] * p[1]);
      S2[2][2] += (p[2] * p[2]);
      S2[0][1] += (p[0] * p[1]);
      S2[0][2] += (p[0] * p[2]);
      S2[1][2] += (p[1] * p[2]);

      if(ps2) // another frame
      {
        const Vec3f& p = (indices) ? ps2[indices[i]] : ps2[i];
        S1 += p;
        S2[0][0] += (p[0] * p[0]);
        S2[1][1] += (p[1] * p[1]);
        S2[2][2] += (p[2] * p[2]);
        S2[0][1] += (p[0] * p[1]);
        S2[0][2] += (p[0] * p[2]);
        S2[1][2] += (p[1] * p[2]);
      }
    }
  }

  int n_points = ((ps2) ? 2 : 1) * ((ts) ? 3 : 1) * n;

  M(0, 0) = S2[0][0] - S1[0]*S1[0] / n_points;
  M(1, 1) = S2[1][1] - S1[1]*S1[1] / n_points;
  M(2, 2) = S2[2][2] - S1[2]*S1[2] / n_points;
  M(0, 1) = S2[0][1] - S1[0]*S1[1] / n_points;
  M(1, 2) = S2[1][2] - S1[1]*S1[2] / n_points;
  M(0, 2) = S2[0][2] - S1[0]*S1[2] / n_points;
  M(1, 0) = M(0, 1);
  M(2, 0) = M(0, 2);
  M(2, 1) = M(1, 2);
}

**Comment by SeanCurtis-TRI** [comment]
Can we assume what that you know what covariance is?

The only *tricky* part of this is the data that you are computing the covariance matrix for. The function is defined in a way to be overloaded. The possible inputs are:

1. A single set of points in R3
2. Two sets of points in R3
3. A set of points defined by a set of triangles
4. Two sets of points defined by a set of triangles (the same triangles at two different positions)
5. A *subset* of triangles at one or two sets of points.

It is the possible permutations of 1-5 which make this not look like a straight forward covariance calculations. But ultimately, you're getting mean values and mean of products you just need to know what samples are going into it.

**Comment by sherm1** [comment]
Wow -- great explanation! That should be in the comments for that code. @betterfan, if @SeanCurtis-TRI's comments were helpful can you submit a PR to add them to the code?

**Comment by betterfan** [comment]
i'm sorry,^_^, what i mean is the covariance matrix calculation of a single set of points in R3,  that means
ps2=NULL, ts=NULL, indices=NULL, then this function may be going like this 
{
for(int i = 0; i < n; ++i)
{
const Vec3f& p = (indices) ? ps[indices[i]] : ps[i];
S1 += p;
S2[0][0] += (p[0] * p[0]);
S2[1][1] += (p[1] * p[1]);
S2[2][2] += (p[2] * p[2]);
S2[0][1] += (p[0] * p[1]);
S2[0][2] += (p[0] * p[2]);
S2[1][2] += (p[1] * p[2]);
}

int n_points = ((ps2) ? 2 : 1) * ((ts) ? 3 : 1) * n;

M(0, 0) = S2[0][0] - S1[0]*S1[0] / n_points;
M(1, 1) = S2[1][1] - S1[1]*S1[1] / n_points;
M(2, 2) = S2[2][2] - S1[2]*S1[2] / n_points;
M(0, 1) = S2[0][1] - S1[0]*S1[1] / n_points;
M(1, 2) = S2[1][2] - S1[1]*S1[2] / n_points;
M(0, 2) = S2[0][2] - S1[0]*S1[2] / n_points;
M(1, 0) = M(0, 1);
M(2, 0) = M(0, 2);
M(2, 1) = M(1, 2);
}
what i mean is the calculation process above is hard to understand, and  the calculation result is different with matlab when using the same data. can you explain clearly? thank you ...^-^ @SeanCurtis-TRI @sherm1

**Comment by SeanCurtis-TRI** [comment]
Ah, I see what you mean. There are a number of possibilities. To help answer your question, could you include the data, the matlab operations you're doing and the results you get in both cases?

**Comment by betterfan** [comment]
yeah, you give a clear explanation, it's really a missing n_points，i have thought  there are some  tricks in the code, haha...^-^ @jmirabel ,do you think so? @SeanCurtis-TRI

**Comment by jmirabel** [comment]
@betterfan If I remember well, the covariance is used to find the principal axes to:
- orient a bounding box of a mesh,
- know on which axis to cut it.
I do not think the (potentially) missing `n` matters for that purpose. Did you have problems with it ?

**Comment by betterfan** [comment]
actually, it doesn't matter the purpose you mentioned, but the result is not a covariance matrix according  to the definition, and why not multiply by n_points, like  this
M(0, 0) = S2[0][0] * n_points- S1[0]*S1[0];

**Comment by betterfan** [comment]
@jmirabel actually, it doesn't matter the purpose you mentioned , but the result is not a covariance matrix according to the definition, and why not multiply by n_points, like this
`M(0, 0) = S2[0][0] * n_points- S1[0]*S1[0];`

**Comment by jmirabel** [comment]
If you have some time, it would be great if you could:
1. check with your matlab code that the error comes from the missing division by `n_points` (BTW, you should divide by `n_points`, not multiply),
2. provide the data, matlab operations and results you get, as @SeanCurtis-TRI mentionned,
4. propose a PR that cleans the code of the function and adds some useful comments.
